### PR TITLE
Show dialogue only when there are unsaved changes

### PR DESCRIPTION
**Pull request by abmyii**
Fix for #17212. Just a POC for now, requires a few changes. Is this an ideal solution - I couldn't find a function that is called on every change so I've got the update counter code in each change function.
Also, would you like me to amend any changes to the commit rather than make new ones?

### CODE REVIEW

**Comment by abmyii** [comment]
If the mods approve then that should be fairly easy to implement. Thanks for that!

**Comment by matjaeck** [comment]
@abmyii teinarss is one of the few developers with push access. Although he didn't do explicitly, he can request changes without needing other people's approval. You're free to disagree with the request, but should explain why in that case.

**Comment by abmyii** [comment]
The last warning causing the checks to fail baffles me slightly. I can't think of a way of getting around it without making the names uppercase which doesn't really make sense. I get this error if I don't declare it as public: `'EditorActionManager.lastSave' is inaccessible due to its protection level`
With my very shallow knowledge of C# I expected it to be implicitly public until explicitly privatised.

**Comment by abmyii** [comment]
Oh I see! I thought his was a suggestion to be discussed with others, not an immediate request. Sorry for the misunderstanding @teinarss!

**Comment by abmyii** [comment]
Done; though I think it would look slightly better if it showed a timestamp, but that would increase the complexity and perhaps require some deeper changes.

**Comment by abmyii** [comment]
Actually it wouldn't - the timestamp could be passed as an argument during initialisation and then concatenated to the `Text` string.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs [hunk 1]
```diff
@@ -355,16 +358,26 @@ void CreateExitEditorButton()
 			if (world.Type != WorldType.Editor)
 				return;
 
+			var actionManager = world.WorldActor.Trait<EditorActionManager>();
 			var button = AddButton("EXIT_EDITOR", "Exit Map Editor");
-			button.OnClick = () =>
```
Comment (teinarss) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: This shouldn't be removed from here.

**Review by teinarss — CHANGES_REQUESTED** [review]

**Comment by abmyii** [comment]
The implementation seems to be complete. The `IsModification` property had to be set to `true` in each class since the default value for a bool is `false` and in interfaces one cannot set a default value for a property. I thought about using `notModification` but that would have been quite confusing for posterity - not good practice.

**Comment by abmyii** [comment]
Any updates?

**Comment by matjaeck** [comment]
It may take several weeks before somebody can review this in detail. Since #17212 was put on the Next+1 milestone (aimed at the release after next release), this PR has lower priority than issues and PR's aimed at Next.

**Comment by abmyii** [comment]
Right, I see. Thanks!

**Comment by teinarss** [comment]
If you open a map and then just exit without making any changes the dialog still shows.

IMO a better solution would to Peek at the undo and see if its a `Opened` or `Saved` event and also check that the redo is empty.

Also i don't see the point in having the timestamp in the saved, but thats is opened for discussion.

**Comment by abmyii** [comment]
I fixed that but I think your suggested method is superior so I will attempt to implement it. As for showing the time saved at, I think it is useful as a reference.

**Comment by abmyii** [comment]
I came across a problem; the EditorActionContainer doesn't contain the text, just an arbitrary id. I'm probably going to add a new status instead.

**Comment by teinarss** [comment]
> 
> 
> I came across a problem; the EditorActionContainer doesn't contain the text, just an arbitrary id. I'm probably going to add a new status instead.

You can do `Peek() is OpenMapAction || Peek() is SaveMapAction`

**Comment by abmyii** [comment]
Oops, didn't mean to do that.
I see, that makes sense - type checking. Didn't think of that.

**Comment by abmyii** [comment]
I get this:
```
The given expression is never of the provided ('OpenMapAction') type
```

**Comment by abmyii** [comment]
With this code:
```
return undoStack.Peek() is OpenMapAction;
```

**Comment by abmyii** [comment]
Should I be using `typeof`?

**Comment by teinarss** [comment]
> 
> 
> With this code:
> 
> ```
> return undoStack.Peek() is OpenMapAction;
> ```

Should be `undoStack.Peek().Action is OpenMapAction`

**Comment by abmyii** [comment]
Duh! Sorry about that.

**Comment by abmyii** [comment]
All right, done! Thanks for all of your help!
As for the timestamp, what do you think we should do about it?

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs [hunk 2]
```diff
@@ -339,9 +339,10 @@ void CreateSaveMapButton()
 			{
 				hideMenu = true;
 				var editorActorLayer = world.WorldActor.Trait<EditorActorLayer>();
+				var actionManager = world.WorldActor.Trait<EditorActionManager>();
 				Ui.OpenWindow("SAVE_MAP_PANEL", new WidgetArgs()
 				{
-					{ "onSave", (Action<string>)(_ => hideMenu = false) },
+					{ "onSave", (Action<string>)(_ => { hideMenu = false; actionManager.Add(new SaveMapAction()); }) },
```
Comment (teinarss) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: Should be on separate line.

File: OpenRA.Mods.Common/Traits/World/EditorActionManager.cs [hunk 3]
```diff
@@ -114,6 +114,11 @@ public void Forward(int id)
 			while (undoStack.Peek().Id != id)
 				Redo();
 		}
+
+		public bool IsModified()
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: I think `HasUnsavedItems` would be better naming.

**Review by teinarss — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/World/EditorActionManager.cs (hunk 3 above)
Comment (abmyii) [reply OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: Done.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs (hunk 2 above)
Comment (abmyii) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: Like so?

Comment (teinarss) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: We can leave it as you had original.

Comment (abmyii) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: I'll undo that then.

Comment (abmyii) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: Done.

File: OpenRA.Mods.Common/Traits/World/EditorActionManager.cs [hunk 4]
```diff
@@ -114,6 +114,11 @@ public void Forward(int id)
 			while (undoStack.Peek().Id != id)
 				Redo();
 		}
+
+		public bool HasUnsavedItems()
+		{
+			return !(undoStack.Peek().Action is OpenMapAction || undoStack.Peek().Action is SaveMapAction);
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: should also check that the `redoStack` is empty

Comment (abmyii) [reply OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: What scenario would that account for?

Comment (teinarss) [reply OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: When undoing some changes back to the Opened or Saved action, At least in Photoshop you will get a prompt warning about unsaved changes.

Comment (abmyii) [reply OpenRA.Mods.Common/Traits/World/EditorActionManager.cs]: Oh right I see. I'll do that now.

**Comment by abmyii** [comment]
One (separate) issue is that saving shouldn't count as an action which prevents redoing, which will also change the `HasUnsavedItems` logic.

**Comment by matjaeck** [comment]
The information at which action a map was saved is useful I think. Can we add "*(saved)*" to the text of the last action instead when saving?

**Comment by abmyii** [comment]
Should it be formatted as `Saved at <timestamp>`?

**Comment by matjaeck** [comment]
I think the timestamp is not needed.

**Comment by abmyii** [comment]
I think it would be useful as a reference for the user. Actually I think this isn't an ideal implementation generally - saving shouldn't even be an action (per se); it should be logged internally and shown at the top amongst the other info labels rather than in the history. Thoughts?

**Comment by teinarss** [comment]
I did check how it's implemented in other programs such as blender,
inkscape,3d studio max and Photoshop. And none of them did show it as an
action. Only Photoshop showed it as notiation on the latest action. E.G
`Added actor (... )  - saved`

**Comment by abmyii** [comment]
> `Added actor (... )  - saved`
Not a bad idea - append it to the previous action's `Text`. Still, I'm uncertain in what way saving and redoing should be related - should saving prevent redoing or not? If not then we cannot include the `HasRedos` check.
My preferred method of showing saving is to have a label denoting the state (`Saved` or `Unsaved`/`Changes made`) at the top (in the "toolbar" area).

**Comment by teinarss** [comment]
It's not flawless in photoshop either, because after a while you will have
several actions with "saved" state. And its not updated when going back and
forth in the history.

Showing the full path to the map with a "*" indicating an unsaved state
could be an alternative fix.

**Comment by abmyii** [comment]
> Showing the full path to the map with a "*" indicating an unsaved state
could be an alternative fix.

That requires the path to be shown (which it currently isn't). What sort of styling should it have?

**Comment by abmyii** [comment]
Made a quick POC.

**Comment by abmyii** [comment]
@teinarss Could you please evaluate the proof-of-concept?

**Comment by abmyii** [comment]
@teinarss Any updates?

**Comment by teinarss** [comment]
Thinking the best would to remove all the visual stuff and only have the
functionality for the preventing showing the dialog if the map is saved.

**Comment by abmyii** [comment]
Right, I'll do that now.

**Comment by abmyii** [comment]
There, I think that works as intended. I'll just refactor a bit and it should be ready.

**Comment by abmyii** [comment]
I had to revert back to the old method of storing state as there is not SaveAction to check for. Is that OK?

**Review by teinarss — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/062bb2b5ddca51fd67c85f76b4792405aff958e1...0ed2b2f7733ea451dcd54daba8c47409e7e012c5)

**Comment by abmyii** [comment]
Awesome, thanks a lot for your assistance @teinarss.

**Pull request merged**